### PR TITLE
Avoid printf for libcxxabi error message reporting

### LINKED ISSUE #11142: libcxxabi depends on printf formatting (+6KB, ~4.6% overhead)
Surprisingly, the following program depends on `printf`:

```c++
class Foo
{
public:
	Foo()
	:counter(5)
	{

	}

	int counter;
};

int testfunc()
{
	static Foo foo;
	return ++foo.counter;
}

int main()
{
	testfunc();
}
```

This is because static initialization of `static Foo foo` will inject dependencies to `cxa_guard_acquire()` and `cxa_guard_release()`, which will have dependencies to cxxabi `abort_message`, which is

```c++
_LIBCXXABI_HIDDEN _LIBCXXABI_NORETURN void
abort_message(const char *format, ...) __attribute__((format(printf, 1, 2)));
```

This will then in turn bring in the full dependency to printf, which is pretty large.

cxxabi has two aborts: `ABORT()` and this printf depending `ABORT_WITH_MESSAGE()`.

It would be great to turn these both in e.g. >= `-O2` and >= `-Os` release builds to instead emit the `emscripten_wasm_trap()` instruction instead (#11141) for minimal code size, and only in debug builds have all the large code size requiring calls.

### PR DESCRIPTION
**Pull request by kripken**
Show those error messages using console.error directly,
avoiding printf. This doesn't show all the format args, but
it is usually fully sufficient, e.g., for an error like

```
__cxa_guard_acquire detected recursive initialization
```

which has no format args anyhow (which seems common).
Also, we get a nice stack trace anyhow in a js/wasm
environment, so these messages can be more useful than
other platforms.

This change avoids linking in vfprintf stdio support just for
this error reporting.

Also implement the direct `emscripten_console_log` etc.
methods in standalone. The test shows that we
avoid all the stdio overhead in non-standalone mode,
which standalone is a nice comparison for (as standalone
must link in some stdio to print - we can't use `console.log`
in JS). The test also shows standalone works (no non-wasi
imports) and avoids printf/formatting etc. at least.

Fixes #11142

### CODE REVIEW

**Comment by juj** [comment]
Would it be possible to instead emit `__builtin_trap()` instead of a print? Having release build code pull in C strings for error messages seems excessive?

`console.error` is much better than `printf` for applications that do not use `printf` otherwise. Though if an application used `printf`, but not `emscripten_console_error`, then this PR actually increases size instead of decreasing it? (since the app will then have deps to both `printf` and `emscripten_console_error`)

I don't think there is much value to have the formatted error messages in release builds, the callstack alone should be suitable information? (one can do a debug build to figure out more details if necessary)

File: tests/other/metadce/libcxxabi_message.cpp [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+  Foo() : counter(5) {}
+  int counter;
+};
+
+int testfunc() {
+  // libcxxabi will emit some checks for proper static initialization, which
+  // can call abort_message()
+  static Foo foo;
+  return ++foo.counter;
+}
```
Comment (sbc100) [inline tests/other/metadce/libcxxabi_message.cpp:12]: Can you explain what this code is doing exactly to cause the abort?  Is there some way would exercise `abort_message` more directly?    (Or simply this code.. at the very least you don't testfunc to be outside of main right?)

Also, should we run this also and check for its output on the console?

File: tests/other/metadce/libcxxabi_message.cpp [hunk 2]
```diff
@@ -0,0 +1,16 @@
+class Foo {
+public:
+  Foo() : counter(5) {}
+  int counter;
+};
```
Comment (sbc100) [inline tests/other/metadce/libcxxabi_message.cpp:5]: could this be written as:

```
// Struct with constructor 
struct Foo { int counter = 5 };
```

File: tests/other/metadce/libcxxabi_message.cpp (hunk 1 above)
Comment (juj) [reply tests/other/metadce/libcxxabi_message.cpp:12]: Oh, this code is not causing an abort, but just compiles in code that contains a reference to the abort. There are lots of ways in libcxxabi that lead to `abort()` or `abort_message()` being called.. Search for `ABORT_WITH_MESSAGE()` -> `abort_message()` -> `printf()`

File: tests/other/metadce/libcxxabi_message.cpp (hunk 2 above)
Comment (kripken) [reply tests/other/metadce/libcxxabi_message.cpp:5]: Sounds good, I can change to that.

File: tests/other/metadce/libcxxabi_message.cpp (hunk 1 above)
Comment (sbc100) [reply tests/other/metadce/libcxxabi_message.cpp:12]: Can you elaborate a little in your comment here @kripken.

Something like: "The point of this test is simply to pull in enough to libc++abi to include its internal `abort_message` code.  We want to ensure that (in release builds) this does not transitively pull in printf formatting code." ?

**Comment by kripken** [comment]
@juj yeah, maybe we can print even less here. We'll have a callstack, but in theory inlining could mess it up. But that would be ok in a debug build, so it seems safe. And I've never even seen these assertions be useful, to be honest.

We don't have a way to make a debug build of system libs, currently. libc++abi is always built with the same options (except for the variations we support, like except/noexcept). That would be nice to add eventually.

File: tests/other/metadce/libcxxabi_message.cpp (hunk 1 above)
Comment (kripken) [reply tests/other/metadce/libcxxabi_message.cpp:13]: Sure, done now.

**Comment by sbc100** [comment]
How about you go with limited "console.error" version of this change .. and we can looking into a possible further optimization (at the cost of maintained more variants of the library), later?

Removing the formatting code is the big win here.

**Comment by kripken** [comment]
I agree it's the bigger win, but I agree with @juj - we may as well do the whole thing. There is essentially no value in these assertions that I am aware of, as I said. Unless you are? (if it's just being cautious, I think that's overly cautious, given that we will definitely have a useful stack trace + function name)

**Comment by sbc100** [comment]
Ok.  In that case can you remove the `standalone_wasm_stdio.c` from this change since its no longer related.

**Comment by kripken** [comment]
Oh right, thanks, done.

File: tests/other/metadce/hello_libcxx_fastcomp_O2_EMULATED_FUNCTION_POINTERS.sent [hunk 3]
```diff
@@ -22,6 +22,7 @@ _fd_write
 _getenv
 _llvm_stackrestore
 _llvm_stacksave
+_llvm_trap
```
Comment (juj) [inline tests/other/metadce/hello_libcxx_fastcomp_O2_EMULATED_FUNCTION_POINTERS.sent:25]: Wait: wasn't this supposed to become a wasm `(unreachable)` opcode? This suggests this is generating a function?

**Comment by juj** [comment]
Related to `abort_message()`, there is the function `abort()` that is used, which might be nice to turn to `(unreachable)` as well. But perhaps that's a separate body of work.

File: tests/other/metadce/hello_libcxx_fastcomp_O2_EMULATED_FUNCTION_POINTERS.sent (hunk 3 above)
Comment (kripken) [reply tests/other/metadce/hello_libcxx_fastcomp_O2_EMULATED_FUNCTION_POINTERS.sent:25]: That's just on fastcomp, which doesn't map it optimally like the LLVM wasm backend does.

**Comment by kripken** [comment]
`abort()` is probably interesting too, yeah. It may be harder as it's used more widely and we'd need to be sure the stack trace would be correct everywhere. Also, `abort()` is just a quick call to JS, and JS needs such a function anyhow, so not sure it would help (but worth exploring).

**Pull request merged**